### PR TITLE
feat(core): Add SAML login setup

### PR DESCRIPTION
**Pull request by flipswitchingmonkey**
Github issue / Community forum post (link here to close automatically):

### CODE REVIEW

File: packages/cli/src/sso/saml/routes/saml.controller.ee.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ */
+
+export const samlController = express.Router();
+samlController.use(samlEnabledMiddleware);
+
+/**
+ * POST /sso/saml/acs
+ * Assertion Consumer Service endpoint
+ */
+samlController.post(SamlUrls.acs, async (req: express.Request, res: express.Response) => {
```
Comment (krynble) [inline packages/cli/src/sso/saml/routes/saml.controller.ee.ts:21]: Some providers might redirect using GET so we should also support this method. I think authentik allows you to change to redirect and then it'll be a GET request.

Screenshot 2023-02-23 at 17 22 26

File: packages/cli/src/sso/saml/samlHelpers.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+export const isSamlPreferences = (candidate: unknown): candidate is SamlPreferences => {
+	const o = candidate as SamlPreferences;
+	return typeof o === 'object' && typeof o.metadata === 'string' && typeof o.mapping === 'object';
+};
+
+export function generatePassword(): string {
+	const length = 18;
+	const charset = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
+	const charsetNoNumbers = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
+	const randomNumber = Math.floor(Math.random());
```
Comment (krynble) [inline packages/cli/src/sso/saml/samlHelpers.ts:33]: This will always yield 0, is this expected?

**Review by krynble — CHANGES_REQUESTED** [review]
Looks good, some minor changes or coments

File: packages/cli/src/config/schema.ts [hunk 3]
```diff
@@ -813,6 +813,12 @@ export const schema = {
 				},
 			},
 		},
+		authenticationMethod: {
+			doc: 'How to authenticate users (e.g. "email", "ldap", "saml")',
+			format: String,
+			default: 'email',
+			env: 'N8N_AUTHENTICATION_METHOD',
```
[the remark this answers is not in the document]
Comment (flipswitchingmonkey) [reply packages/cli/src/config/schema.ts:820]: sure. we can store it in the db instead

File: packages/cli/src/sso/saml/samlHelpers.ts (hunk 2 above)
Comment (flipswitchingmonkey) [reply packages/cli/src/sso/saml/samlHelpers.ts:33]: lol, no, good catch

File: packages/cli/src/sso/saml/routes/saml.controller.ee.ts (hunk 1 above)
Comment (flipswitchingmonkey) [reply packages/cli/src/sso/saml/routes/saml.controller.ee.ts]: updated and sort of works, but ends with  ERR_MISSING_SIG_ALG issue. This may be an authentik setup issue though

File: packages/cli/src/sso/saml/middleware/samlEnabledMiddleware.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+	console.log(isSamlEnabled(), req.user?.globalRole.name);
+	if (isSamlLicensed() && req.user?.globalRole.name === 'owner') {
+		next();
+	} else {
+		res.status(401).json({ status: 'error', message: 'Unauthorized' });
+	}
+};
+
+export const samlLicensedAndEnabledMiddleware: RequestHandler = (req, res, next) => {
+	console.log(isSamlEnabled(), isSamlLicensed(), isSamlCurrentAuthenticationMethod());
```
Comment (krynble) [inline packages/cli/src/sso/saml/middleware/samlEnabledMiddleware.ts:20]:
```suggestion
```

File: packages/cli/src/sso/saml/saml.service.ee.ts [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+	): Promise<
+		| {
+				authenticatedUser: User | undefined;
+				attributes: SamlUserAttributes;
+				onboardingRequired: boolean;
+		  }
+		| undefined
+	> {
+		const attributes = await this.getAttributesFromLoginResponse(req, binding);
+		if (attributes.email) {
```
Comment (krynble) [inline packages/cli/src/sso/saml/saml.service.ee.ts:104]: Maybe not for this iteration, but if email cannot be find as part of the mapping, we need to warn the user about it.

File: packages/cli/src/ResponseHelper.ts [hunk 6]
```diff
@@ -190,7 +190,7 @@ export function send<T, R extends Request, S extends Response>(
 		try {
 			const data = await processFunction(req, res);
 
-			sendSuccessResponse(res, data, raw);
+			if (!res.headersSent) sendSuccessResponse(res, data, raw);
```
Comment (krynble) [inline packages/cli/src/ResponseHelper.ts:193]: Is this change necessary since saml-related controllers are not using the annotations?

**Review by krynble — CHANGES_REQUESTED** [review]

File: packages/cli/src/ResponseHelper.ts (hunk 6 above)
Comment (flipswitchingmonkey) [reply packages/cli/src/ResponseHelper.ts:193]: yes it's for the default auth /login endpoint to be able to redirect to the initsso endpoint

**Review by krynble — DISMISSED** [review]

**Review by krynble — APPROVED** [review]
LGTM

**Comment by janober** [comment]
Got released with [`n8n@0.217.2`]((https://github.com/n8n-io/n8n/releases/tag/n8n%400.217.2))
